### ISSUE TITLE
Need to pull sampled data for many tags at the same time

### ISSUE THREAD

**Issue by KHDOUDI** [Hugovdberg/PIconnect#526]
Hello,

Thank you very much for this very well made package, it's really very helpful.

I am struggling since some weeks on how to write a code to pull sampled data "500 ms" for many tags at the same time. i have about 120 tag names which i want to pull data in parallel as numpy array.

some tag names from my 120 list:

KE_91_T_EX_Motors_SHUTTLE_UDFB_MOTOR_TEMP.PV
KE_91_T_EX_Motors_SHUTTLE_UDOP_MOTOR_TEMP.PV
KE_91_T_EX_Motors_SHUTTLE_FBBK_MOTOR_TEMP.PV
KE_91_T_EX_Motors_SHUTTLE_FBOP_MOTOR_TEMP.PV
KE_91_T_EX_Motors_BLAST_UPBK_MOTOR_TEMP.PV
KE_91_T_EX_Motors_BLAST_UPOP_MOTOR_TEMP.PV
KE_91_T_EX_Motors_BLAST_DNBK_MOTOR_TEMP.PV
KE_91_T_EX_Motors_BLAST_DNOP_MOTOR_TEMP.PV

My actual script (which is not working correctly) is below:

```python
import csv
import pandas as pd
import numpy as np
import PIconnect as PI

with open("tags.txt") as file_in:
    lines = []
    for line in file_in:
        lines.append(line)

f=open('all_data.txt','a')

with PI.PIServer() as server:
    print(server.server_name)
    for i in range(8):
        points = server.search(lines)[i]
        data = points.interpolated_values('*-1h', '*', '10m')
        print(data)
        x=data.to_frame()
        np.savetxt(f, data, '%1.3f', newline=", ")
        f.write("\n")
```

Thank you for your support

**Comment by Hugovdberg** [comment]
What is currently the problem with the script you have? My guess would be that the data are not aligned at the same timestamps because the relative time `'*-1h'` is evaluated for each tag independently, and therefore is shifted by the amount of time it took to download the previous data. 

This is something I'm still hoping to fix sometime by implementing the `PIPointList` class, that allows you to query multiple PIPoints at the same time. This require some rewriting of the current code for which I currently do not have the time.

What you could do for now is calculate the start time in advance and pass that to your query (I took the liberty to clean up your code a little):

```python
import datetime

import numpy as np
import pandas as pd
import PIconnect as PI

# Converting a file handle to list automatically splits by line
with open("tags.txt") as file_in:
    tags = list(file_in)

# Subtract 1 hour from the current time and convert to string
start_time = datetime.datetime.now() - datetime.timedelta(hours=1)
start_time = start_time.strftime("%Y-%m-%d %H:%M:%S.%f")

with PI.PIServer() as server:
    # You can pass a list of tags and get a list of points back
    points = server.search(tags)
    # Create a list of PISeries for each point in the result
    data = [point.interpolated_values(start_time, "*", "10m") for point in points]
    # Concatenate all the series to a single dataframe,
    # axis=1 means make each Series a column
    data = pd.concat(data, axis=1)
# Write the entire dataframe at once to a file
data.to_csv('all_data.txt')
```

Hope this is of any use to you, if not please clarify what's going wrong :-)

**Comment by KHDOUDI** [comment]
Dear Hugo,

i am really happy to test your code because it's works very well as i am expecting. for that i thank you very much for your support. you just solved my problem :-)

Have a nice day.

**Comment by Hugovdberg** [comment]
Closing this issue for now as it is actually a duplicate of #74. Will add a backreference so we test whether selecting from a list of points actually use the same timestamps in the query.
